src/network/stream_reader.rs

```diff
@@ -0,0 +1,338 @@
+//
+// To the extent possible under law, the author(s) have dedicated all
+// copyright and related and neighboring rights to this software to
+// the public domain worldwide. This software is distributed without
+// any warranty.
+//
+// You should have received a copy of the CC0 Public Domain Dedication
+// along with this software.
+// If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
+//
+
+//! Stream reader
+//!
+//! This module defines `StreamReader` struct and its implementation which is used
+//! for parsing incoming stream into separate `RawNetworkMessage`s, handling assembling
+//! messages from multiple packets or dealing with partial or multiple messages in the stream
+//! (like can happen with reading from TCP socket)
+//!
+
+use std::fmt;
+use std::io;
+use std::io::Read;
+use std::sync::mpsc::Sender;
+
+use util;
+use network::message::{NetworkMessage, RawNetworkMessage};
+use consensus::encode;
+
+/// A response from the peer-connected socket
+pub enum SocketResponse {
+    /// A message was received
+    MessageReceived(NetworkMessage),
+    /// An error occurred and the socket needs to close
+    ConnectionFailed(util::Error, Sender<()>)
+}
+
+/// Struct used to configure stream reader function
+pub struct StreamReader<'a> {
+    /// Size of allocated buffer for a single read opetaion
+    pub buffer_size: usize,
+    /// Stream to read from
+    pub stream: &'a mut Read,
+    /// Buffer containing unparsed message part
+    unparsed: Vec<u8>
+}
+
+impl<'a> fmt::Debug for StreamReader<'a> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "StreamReader with buffer_size={} and unparsed content {:?}",
+               self.buffer_size, self.unparsed)
+    }
+}
+
+impl<'a> StreamReader<'a> {
+    /// Constructs new stream reader for a given input stream `stream` with
+    /// optional parameter `buffer_size` determining reading buffer size
+    pub fn new(stream: &mut Read, buffer_size: Option<usize>) -> StreamReader {
+        StreamReader {
+            stream,
+            buffer_size: buffer_size.unwrap_or(64 * 1024),
+            unparsed: vec![]
+        }
+    }
+
+    /// Reads stream and parses messages from its current input,
+    /// also taking into account previously unparsed partial message (if there was such).
+    pub fn read_messages(&mut self) -> Result<Vec<RawNetworkMessage>, encode::Error> {
+        let mut messages: Vec<RawNetworkMessage> = vec![];
+        let mut data = vec![0u8; self.buffer_size];
+        while messages.len() == 0 {
+            let count = self.stream.read(&mut data)?;
+            if count > 0 {
+                self.unparsed.extend(data[0..count].iter());
+            }
+            messages.append(&mut self.parse()?);
+        }
+        return Ok(messages)
+    }
+
+    fn parse(&mut self) -> Result<Vec<RawNetworkMessage>, encode::Error> {
+        let mut messages: Vec<RawNetworkMessage> = vec![];
+        while self.unparsed.len() > 0 {
```

Comment:
change 
```
while self.unparsed.len() > 0 {
```
to
```
if self.unparsed.len() > 0 {
```
and messages will be delivered one by one. I do not see how this would lock anything.